src/python/py/models/builder.py

```diff
@@ -1074,25 +1143,127 @@ def make_mlp_fc(self, layer_id, mlp, root_input):
+        self.make_matmul(self.make_shard(mlp.fc2.weight, sharding_axis=1), fc2_matmul_name, root_input=f"{act_fn_name}/output_0")
+
+        if self.world_size > 1:
+            all_reduce_name = f"/model/layers.{layer_id}/mlp/all_reduce"
+            self.make_all_reduce(all_reduce_name, f"{fc2_matmul_name}/output_0")
+            fc2_matmul_name = all_reduce_name
+
         fc2_add_name = f"/model/layers.{layer_id}/mlp/fc2/Add"
         self.make_add_bias(mlp.fc2.bias.detach().numpy(), fc2_add_name, root_input=f"{fc2_matmul_name}/output_0")
 
         # Assign output 0 of MLP layer as output of last layer
         self.mlp_attrs["output_0"] = f"{fc2_add_name}/output_0"
 
+    def make_block_sparse_moe(self, layer_id, bsm, root_input):
+        if bsm is None:
+            return
+
+        num_experts = self.moe_attrs["num_experts"]
+        top_k = self.moe_attrs["top_k"]
+        activation_type = self.moe_attrs["activation_type"]
+        normalize_routing_weights = self.moe_attrs["normalize_routing_weights"]
+        
+        # Make MoE nodes
+        gate_name = f"/model/layers.{layer_id}/moe/gate/MatMul"
+        self.make_matmul(bsm.gate.weight.detach().numpy(), gate_name, root_input)
+
+        shape_name = f"/model/layers.{layer_id}/moe/gate/Shape"
+        self.make_shape(shape_name, f"{gate_name}/output_0", shape=[3])
+
+        gather_name = f"/model/layers.{layer_id}/moe/gate/Gather"
+        self.make_gather(gather_name, [f"{shape_name}/output_0", "/model/constants/TensorProto.INT64/0D/2"], axis=0)
+
+        unsqueeze_name = f"/model/layers.{layer_id}/moe/gate/Unsqueeze"
+        self.make_unsqueeze(unsqueeze_name, [f"{gather_name}/output_0", "/model/constants/TensorProto.INT64/1D/0"], dtype=TensorProto.INT64, shape=[1])
+
+        concat_name = f"/model/layers.{layer_id}/moe/gate/Concat"
+        self.make_concat(concat_name, ["/model/constants/TensorProto.INT64/1D/-1", f"{unsqueeze_name}/output_0"], dtype=TensorProto.INT64, shape=[2], axis=0)
+
+        gate_reshape_name = f"/model/layers.{layer_id}/moe/gate/Reshape"
+        self.make_reshape(gate_reshape_name, [f"{gate_name}/output_0", f"{concat_name}/output_0"], dtype=self.io_dtype, shape=['num_rows', num_experts])
+
+        moe_name = f"/model/layers.{layer_id}/moe"
+        w1_list = []
+        w2_list = []
+        w3_list = []
+
+        hidden_size = self.hidden_size
+        inter_size = self.intermediate_size
+
+        for i in range(num_experts):
+            w1_list.append(torch.reshape(bsm.experts[i].w1.weight, (hidden_size, inter_size)))
+            w2_list.append(torch.reshape(bsm.experts[i].w2.weight, (inter_size, hidden_size)))
+            w3_list.append(torch.reshape(bsm.experts[i].w3.weight, (hidden_size, inter_size)))
+
+        moe_expert_1_name = f"model.layers.{layer_id}.moe.weight_1"
+        moe_expert_2_name = f"model.layers.{layer_id}.moe.weight_2"
```

Comment:
Can the expert weight names be named as `model.layers.{layer_id}.moe.experts.{expert_num}.weight` instead?